src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -302,62 +298,35 @@ public SNIHandle CreateConnectionHandle(object callbackObject, string fullServer
-                if ((serverNameWithOutProtocol = GetServerNameWithOutProtocol(fullServerName, TdsEnums.TCP + ":")) != null)
-                {
-                    sniHandle = CreateTcpHandle(serverNameWithOutProtocol, timerExpire, callbackObject, parallel, ref spnBuffer, isIntegratedSecurity);
-                }
-                // when np protocol is specified
-                else if ((serverNameWithOutProtocol = GetServerNameWithOutProtocol(fullServerName, TdsEnums.NP + ":")) != null)
+                switch (details.ConnectionProtocol)
                 {
-                    sniHandle = CreateNpHandle(serverNameWithOutProtocol, timerExpire, callbackObject, parallel);
-                }
-                // possibly error case
-                else
-                {
-                    int portOrInstanceNameIndex = Math.Max(fullServerName.LastIndexOf(','), fullServerName.LastIndexOf('\\'));
-                    string serverNameWithOutPortOrInstanceName = portOrInstanceNameIndex > 0 ? fullServerName.Substring(0, portOrInstanceNameIndex) : fullServerName;
-                    IPAddress address = null;
-
-                    // when no protocol is specified, and fullServerName is IPv6
-                    if (IPAddress.TryParse(serverNameWithOutPortOrInstanceName, out address) && address.AddressFamily == AddressFamily.InterNetworkV6)
-                    {
-                        // default to using tcp if no protocol is provided
-                        sniHandle = CreateTcpHandle(fullServerName, timerExpire, callbackObject, parallel, ref spnBuffer, isIntegratedSecurity);
-                    }
-                    // error case for sure
-                    else
-                    {
-                        // when invalid protocol is specified
-                        if (IsOccursOnce(fullServerName, ':'))
-                        {
-                            SNICommon.ReportSNIError(
-                                SNIProviders.INVALID_PROV, 0,
-                                (uint)(parallel ? SNICommon.MultiSubnetFailoverWithNonTcpProtocol : SNICommon.ProtocolNotSupportedError),
-                                string.Empty);
-                        }
-                        // when fullServerName is in invalid format
-                        else
-                        {
-                            SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.INVALID_PROV, 0, SNICommon.InvalidConnStringError, string.Empty);
-                        }
-                    }
+                    case DataSource.Protocol.TCP:
+                        sniHandle = CreateTcpHandle(details, timerExpire, callbackObject, parallel, ref spnBuffer, isIntegratedSecurity);
+                        break;
+                    case DataSource.Protocol.NP:
+                        sniHandle = CreateNpHandle(details, timerExpire, callbackObject, parallel);
+                        break;
```

Comment:
Should there be a default case as well?  If there are no other possible options, maybe slightly future proof it with something like:
```C#
default:
    Debug.Fail($"Unexpected connection protocol: {details.ConnectionProtocol}");
    break;
```